src/v/raft/heartbeats.h

```diff
@@ -0,0 +1,460 @@
+        using serde::read_async_nested;
+        using serde::read_nested;
+        c._reminder_buffer = read_nested<row_t>(in, bytes_left_limit);
+        auto initial = read_nested<field_t>(in, bytes_left_limit);
+        auto cnt = read_nested<uint32_t>(in, bytes_left_limit);
+        auto last = read_nested<field_t>(in, bytes_left_limit);
+        auto buffer = read_nested<iobuf>(in, bytes_left_limit);
+        c._data = encoder_t(initial, cnt, last, std::move(buffer));
+        c._cnt = read_nested<size_t>(in, bytes_left_limit);
+    }
+
+    friend inline void
+    write(iobuf& out, deltafor_column<ValueT, DeltaT, FieldT> state) {
+        using serde::write;
+        write(out, state._reminder_buffer);
+        write(out, state._data.get_initial_value());
+        write(out, state._data.get_row_count());
+        write(out, state._data.get_last_value());
+        write(out, state._data.share());
+        write(out, state._cnt);
+    }
+
+private:
+    row_t _reminder_buffer{};
+    encoder_t _data;
+    size_t _cnt{0};
+};
+
+template<typename Func, typename... C>
+void for_each_column(Func f, const C&... values) {
+    auto cursors = std::make_tuple(values.make_cursor()...);
+
+    for (size_t i = 0; i < std::get<0>(cursors).size(); ++i) {
+        auto values = std::apply(
+          [](auto&... c) { return std::make_tuple(c()...); }, cursors);
+        std::apply(f, values);
+        std::apply(
+          [](auto&... c) { return std::make_tuple(c.advance()...); }, cursors);
+    }
+}
+
+struct heartbeat_request_data {
+    model::revision_id source_revision;
+    model::revision_id target_revision;
+    model::offset commit_index;
+    model::term_id term;
+    model::offset prev_log_index;
+    model::term_id prev_log_term;
+    model::offset last_visible_index;
+
+    friend inline void read_nested(
+      iobuf_parser& in,
+      heartbeat_request_data& state,
+      size_t const bytes_left_limit) {
+        using serde::read_nested;
+        read_nested(in, state.source_revision, bytes_left_limit);
+        read_nested(in, state.target_revision, bytes_left_limit);
+        read_nested(in, state.commit_index, bytes_left_limit);
+        read_nested(in, state.term, bytes_left_limit);
+        read_nested(in, state.prev_log_index, bytes_left_limit);
+        read_nested(in, state.prev_log_term, bytes_left_limit);
+        read_nested(in, state.last_visible_index, bytes_left_limit);
+    }
+    friend inline void write(iobuf& out, heartbeat_request_data state) {
+        using serde::write;
+        write(out, state.source_revision);
+        write(out, state.target_revision);
+        write(out, state.commit_index);
+        write(out, state.term);
+        write(out, state.prev_log_index);
+        write(out, state.prev_log_term);
+        write(out, state.last_visible_index);
+    }
+
+    friend bool
+    operator==(const heartbeat_request_data&, const heartbeat_request_data&)
+      = default;
+};
+
+struct heartbeat_reply_data {
```

Comment:
i didn't make them envelopes as it would require adding envelope to every heartbeat data in potentially large `ss::chunked_fifo` of full heartbeats. Will an envelope of a top level request be enough ?

Reply:
For coordinated recovery followers need to request/allow recovery for individual partitions via heartbeats, so we'll need to add a flag to the heartbeat struct (potentially more than a flag - e.g. number of allowed recovery bytes). This can be done with a top-level envelope (by adding a "column" on the side), but is awkward.